### PR TITLE
Updates to train/learn subpackage

### LINKED ISSUE #186: use HuggingFace models in MapReader
**Is your feature request related to a problem? Please describe.**
We would like people to be able to use HuggingFace models in the mapreader pipeline. 

This could either be:
- Add a page in the docs explaining HOW to do this (relatively low effort to do this)
- Write a function/method into the MR code that allows people to do this in a very simple way (higher effort but potentially more useful for non-code-y people).

https://huggingface.co/davanstrien/autotrain-mapreader-5000-40830105612/tree/main would be a good one to use for testing.

**Describe the solution you'd like**
- [x] Have a go at using this model in a new run through of the mapreader pipeline
- [x] Write how to load hugging face models into documentation

Split to #192 for additional things to do

### PR DESCRIPTION
**Pull request by rwood-97**
### Summary
General updates to train/learn subpackage.
Fixes #186 

### Describe your changes
- `AnnotationsLoader` is now part of this subpackage and now creates the datasets and dataloaders
- `ClassifierContainer` MUST be loaded with a model, dataloaders and labels_map
- PEP8 style guide updates
- [x] Update context parts
- [x] Allow hugging face models (transformers) - #186 
- [x] Update docs
- [x] Rename to ``classify``

### Checklist before assigning a reviewer (update as needed)
  
- [x] Self-review code
- [x] Ensure submission passes current tests
- [x] Add tests
  
### Reviewer checklist
  
Please add anything you want reviewers to specifically focus/comment on.

- [ ] Everything looks ok?

### CODE REVIEW

File: mapreader/classify/classifier.py [hunk 1]
```diff
@@ -1507,13 +1431,11 @@ def initialize_model(
             input_size = 299
 
         else:
-            raise ValueError("Invalid model name, exiting...")
+            raise NotImplementedError("[ERROR] Invalid model name.")  # CHECK THIS
```
Comment (kallewesterling) [inline mapreader/classify/classifier.py:1434]: Suggesting a TODO here to make sure we catch it in the future

File: mapreader/classify/classifier.py [hunk 2]
```diff
@@ -1851,37 +1817,37 @@ def load(
     def _print_colors(self):
         """Private function, setting color attributes on the object."""
         # color
-        self.color_lgrey = "\033[1;90m"
-        self.color_grey = "\033[90m"  # broing information
-        self.color_yellow = "\033[93m"  # FYI
-        self.color_orange = "\033[0;33m"  # Warning
+        self.__color_lgrey = "\033[1;90m"
+        self.__color_grey = "\033[90m"  # broing information
```
Comment (kallewesterling) [inline mapreader/classify/classifier.py:1821]: broing or boring?

File: mapreader/classify/classifier_context.py [hunk 3]
```diff
@@ -251,16 +258,33 @@ def train_core(
...
+                                ):
+                                    try:
+                                        outputs = outputs.logits
+                                        aux_outputs = aux_outputs.logits
+                                    except AttributeError as err:
+                                        raise AttributeError(err.message)
+
+                                loss1 = self.criterion(outputs, label_indices)
+                                loss2 = self.criterion(aux_outputs, label_indices)
                                 # XXX From https://discuss.pytorch.org/t/how-to-optimize-inception-model-with-auxiliary-classifiers/7958 # noqa
```
Comment (kallewesterling) [inline mapreader/classify/classifier_context.py:274]: Should we move this to a note in the docstring for the method you think?

File: mapreader/classify/datasets.py [hunk 4]
```diff
@@ -122,23 +176,35 @@ def __getitem__(self, idx: Union[int, torch.Tensor]) -> Tuple[torch.Tensor, Any]
         Returns
         -------
         tuple
```
Comment (kallewesterling) [inline mapreader/classify/datasets.py:178]: might be helpful to add what the tuple contains here - like the type hinting above `Tuple[torch.Tensor, Any]`.. that way, it'll be included when generating the reference docs for the web. (This applies to any Returns sections where applicable.)

File: mapreader/classify/datasets.py [hunk 5]
```diff
@@ -600,20 +701,20 @@ def plot_sample(self, indx: int) -> None:
...
-        plt.imshow(transforms.ToPILImage()(self.__getitem__(indx)[1]))
+        plt.imshow(transforms.ToPILImage()(self.__getitem__(idx)[1]))
         plt.title("Context", size=18)
         plt.xticks([])
         plt.yticks([])
         plt.subplot(1, 2, 2)
         plt.show()
 
-    def __getitem__(self, idx: int) -> Tuple[torch.Tensor, torch.Tensor, Any]:
+    def __getitem__(self, idx: Union[int, torch.Tensor]) -> tuple:
```
Comment (kallewesterling) [inline mapreader/classify/datasets.py:717]: Perhaps add the type hinting in here, from the Returns section below? Makes the codebase so easily usable in VS code and other environs.

**Review by kallewesterling — CHANGES_REQUESTED** [review]
I've read through the code now, but haven't test-run it. (Not sure I'll have time to do that, tbf.. Do you need it?)

File: mapreader/classify/classifier.py (hunk 1 above)
Comment (rwood-97) [reply mapreader/classify/classifier.py:1434]: I actually think I could remove this comment - wanted to check whether passing "resnet18" vs "resnet50" would give you the same model as output but am now confident it doesnt

File: mapreader/classify/datasets.py (hunk 4 above)
Comment (rwood-97) [reply mapreader/classify/datasets.py:178]: yes agree - have updated this and the one in PatchContextDataset

**Comment by kallewesterling** [comment]
Great, thanks!

File: mapreader/annotate/utils.py [hunk 6]
```diff
@@ -664,31 +659,32 @@ def save_annotation(
...
                 new_labels += 1
 
     if len(image_df) > 0:
-        image_df = image_df.set_index("image_id")
-        image_df.to_csv(annot_file, mode="w")
+        #image_df = image_df.set_index("image_id")
+        image_df.to_csv(annot_file, mode="w", sep="\t")
         print(f"[INFO] Save {newly_annotated} new annotations to {annot_file}")
         print(f"[INFO] {new_labels} labels were not already stored")
         print(f"[INFO] Total number of annotations: {len(image_df)}")
```
Comment (kmcdono2) [inline mapreader/annotate/utils.py:690]: "**Saved** {newly_annotated} new annotations to {annot_file}

"Total number of annotations" --> "Total number of **saved** annotations"

**Review by kmcdono2 — CHANGES_REQUESTED** [review]
Looks great! I just made a few comments mostly for clarity to prose.

**Comment by rwood-97** [comment]
@kmcdono2 Thanks, have added in all these changes :)

**Pull request merged**